File: simulator/src/piwarssim/engine/challenges/Challenges.py
from enum import Enum

from piwarssim.engine.challenges.PiNoonChallenge import PiNoonChallenge
from piwarssim.engine.challenges.BlastOffChallenge import BlastOffChallenge
from piwarssim.engine.challenges.CanyonsOfMarsChallenge import CanyonsOfMarsChallenge
from piwarssim.engine.challenges.EcoDisasterChallenge import EcoDisasterChallenge
from piwarssim.engine.challenges.MineSweeperChallenge import MineSweeperChallenge
from piwarssim.engine.challenges.StraightLineSpeedTestChallenge import StraightLineSpeedTestChallenge


def raise_not_implemented():
    raise NotImplemented


class Challenges(Enum):
    PiNoon = (lambda : PiNoonChallenge(),)
    EcoDisaster = (lambda : EcoDisasterChallenge(),)
    CanyonsOfMars = (lambda : CanyonsOfMarsChallenge(),)
    StraightLineSpeedTest = (lambda : StraightLineSpeedTestChallenge(),)
    BlastOff = (lambda : BlastOffChallenge(),)
    MineSweeper = (lambda : MineSweeperChallenge(),)

    def __new__(cls, new_object_function):
        value = len(cls.__members__)
        obj = object.__new__(cls)
        obj._value_ = value
        return obj

    def __init__(self, new_object_function):
        self._new_object_function = new_object_function

    def ordinal(self):
        return self.value

    def new_object(self):
        return self._new_object_function()

    @staticmethod
    def from_ordinal(ordinal):
        for enum_obj in Challenges:
            if enum_obj.value == ordinal:
                return enum_obj

    @staticmethod
    def from_name(name):
        for enum_obj in Challenges:
            if enum_obj.name == name:
                return enum_obj

File: server/piwars-common/src/main/java/org/ah/piwars/virtualrover/game/challenge/Challenges.java
package org.ah.piwars.virtualrover.game.challenge;

import org.ah.piwars.virtualrover.game.PiWarsGame;

public class Challenges {

    private Challenges() { }

    public static Challenge createChallenge(PiWarsGame game, String name) {
        if ("PiNoon".equals(name)) {
            return new PiNoonChallenge(game, name);
        } else if ("EcoDisaster".equals(name)) {
            return new EcoDisasterChallenge(game, name);
        } else if ("CanyonsOfMars".equals(name)) {
            return new CanyonsOfMarsChallenge(game, name);
        } else if ("StraightLineSpeedTest".equals(name)) {
            return new StraightLineSpeedTestChallenge(game, name);
        } else if ("BlastOff".equals(name)) {
            return new BlastOffChallenge(game, name);
        } else if ("MineSweeper".equals(name)) {
            return new MineSweeperChallenge(game, name);
        }

        throw new IllegalArgumentException(name + " not known");
    }
}
